src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,228 @@
+NOTE: mutates tensor!")
+  (:method ((function function) (tensor abstract-tensor))
+    (map-indexes
+     (shape tensor)
+     (lambda (&rest dims)
+       (apply #'(setf tref) (funcall function (apply #'tref tensor dims)) tensor dims)))
+    tensor))
+
+(defgeneric into! (function abstract-tensor)
+  (:documentation "Map indices to elements
+NOTE: mutates tensor!")
+  (:method ((function function) (tensor abstract-tensor))
+    (map-indexes
+     (shape tensor)
+     (lambda (&rest dims)
+       (apply #'(setf tref) (apply function dims) tensor dims)))
+    tensor))
+
+(defgeneric foreach (function abstract-tensor)
+  (:documentation "Call a function with each element of a tensor")
+  (:method ((function function) (tensor abstract-tensor))
+    (map-indexes
+     (shape tensor)
+     (lambda (&rest dims)
+       (funcall function (apply #'tref tensor dims))))
+    tensor))
+
+(defgeneric map-to (function source target)
+  (:documentation "Map from one tensor to another using a given function")
+  (:method ((function function) (source abstract-tensor) (target abstract-tensor))
+    (assert (equalp (shape source) (shape target))
+            () "Incompatible shapes. Cannot map tensor of shape ~a to tensor of shape ~a."
+            (shape source) (shape target))
+    (map-indexes
+     (shape source)
+     (lambda (&rest dims)
+       (apply #'(setf tref) (funcall function (apply #'tref source dims)) target dims)))))
+
+(defgeneric map (function abstract-tensor)
+  (:documentation "Map elements of a tensor")
+  (:method ((function function) (tensor abstract-tensor))
+    (let ((target (copy-tensor tensor)))
+      (map-to function tensor target)
+      target)))
+
+;; TODO: Determine if this makes sense
+(defgeneric into (function abstract-tensor)
+  (:documentation "Map indices to elements")
+  (:method ((function function) (tensor abstract-tensor))
+    (let ((target (copy-tensor tensor)))
+      (into! function target)
+      target)))
+
+(defgeneric sum (abstract-tensor)
+  (:documentation "Get the sum of the elements of a tensor")
+  (:method ((tensor abstract-tensor))
+    (let ((sum 0)) ;; TODO: Make sure this works for all types
+      (foreach (lambda (x) (incf sum x)) tensor)
+      sum)))
+
+(defgeneric scale (abstract-tensor factor)
+  (:documentation "Scale the tensor by a given factor, returning a new tensor")
+  (:method ((tensor abstract-tensor) (factor number))
+    (let ((target (copy-tensor tensor)))
+      (map-to (lambda (x) (* x factor)) tensor target)
+      target)))
+
+(defgeneric scale! (abstract-tensor factor)
+  (:documentation "Scale the tensor by the given factor, storing back into the tensor")
+  (:method ((tensor abstract-tensor) (factor number))
+    (map! (lambda (x) (* x factor)) tensor)))
+
+;; TODO: make sure this doesn't destroy the vector type
+;; TODO: Allow for any number of tensors. Might need to write something to remove keywords from rest
+(defgeneric + (source1 source2 &key target)
```

Comment:
I suggest replacing `&key` by `&optional`, as it takes less time to process. I can anticipate we'll be calling this method within loops where that could make a difference.

Same for all the other generic functions having `target` as a keyword argument.